documentation/modules/upgrading/con-upgrade-listeners.adoc

```diff
@@ -0,0 +1,63 @@
+// This module is included in the following assemblies:
+//
+// upgrading/assembly_upgrade-kafka-versions.adoc
+
+[id='con-upgrade-listeners-{context}']
+= Updating listener configuration
+
+Strimzi provides a `GenericKafkaListener` schema for the configuration of Kafka listeners in a `Kafka` resource.
+
+`GenericKafkaListener` replaces the `KafkaListeners` schema, which is deprecated.
+
+The `GenericKafkaListener` schema is backwards compatible.
+
+With the `GenericKafkaListener` schema, you can configure as many listeners as required,
+as long as their names and ports are unique.
+The listener configuration is defined as an array.
+
+For clients inside the Kubernetes cluster, you can create `plain` (without encryption) or `tls` _internal_ listeners.
+
+For clients outside the Kubernetes cluster, you create _external_ listeners and specify a connection mechanism,
+which can be `nodeport`, `loadbalancer`, `ingress` or `route`.
+
+The `KafkaListeners` schema uses sub-properties for `plain`, `tls` and `external` listeners, with fixed ports for each.
+As part of your upgrade, you can convert listeners configured using the `KafkaListeners` schema into the format of the `GenericKafkaListener` schema.
+
+For example, if you are currently using the following configuration in your `Kafka` configuration:
+
+[source,shell,subs="+quotes,attributes"]
+----
+listeners:
+  plain:
+    # ...
+  tls:
+    # ...
+  external:
+    type: loadbalancer
+    # ...
+----
+
+Convert the listeners into the new format using:
```

Comment:
Can we please stress out two points:
* The name and port numbers here are really important for the backwards compatibility. Using other names will possibly result in differently named listeners in Kafka or services in Kubernetes which will cause problems.
* The options in `configuration` or `overrides` of the old listeners need to be adjusted into the new `configuration format`

Reply:
Updated. I was going to add more of the change details to the release notes, but I think it's useful here too.
So:

> Make sure to use the the *exact* names and port numbers shown here.
> 
> For any `configuration` or `overrides` properties used with the old format, you need to update them to the new format.
> 
> Changes introduced to the listener `configuration`:
> 
> * `overrides` property is merged with the `configuration` section
> * `dnsAnnotations` property has been renamed `annotations`
> * `preferredAddressType` has been renamed `preferredNodePortAddressType`
> * `address` property has been renamed `alternativenames`
> * `loadBalancerSourceRanges` and `externalTrafficPolicy` properties move to the listener configuration from the now deprecated `template`
> 
> For example, this configuration:
> 
> ```
> listeners:
>   external:
>     type: loadbalancer
>     authentication:
>       type: tls
>     overrides:
>       bootstrap:
>         dnsAnnotations:
>           #...
> ````
> 
> Changes to:
> 
>````
> ----
> listeners:
>     #...
>   - name: external
>     port: 9094
>     type:loadbalancer
>     tls: true
>     configuration:
>       bootstrap:
>         annotations:
>           #...
> ````
> 
> **Important**
> The name and port numbers shown in the new configuration *must* be used for backwards compatibility.
> Using any other names will cause problems with the naming of Kafka listeners and Kubernetes services.